src/keys/msgsig.rs

```diff
@@ -0,0 +1,128 @@
+
+/// Signing arbitrary messages and verify message signatures
+
+use bitcoin::hashes::Hash;
+use bitcoin::secp256k1::recovery::RecoverableSignature;
+use bitcoin::secp256k1::{All, Message, Secp256k1, SecretKey};
+use bitcoin::util::misc::{signed_msg_hash, MessageSignature};
+use bitcoin::{Address, Network, PrivateKey, PublicKey};
+
+/// Trait for message signers
+pub trait MessageSigner<S> {
+    /// Sign an arbitrary message
+    fn sign(&self, msg: &str) -> S;
+}
+
+/// Trait for message signature verifiers
+pub trait MessageSignatureVerifier<S> {
+    /// Verify a signature against an arbitrary message
+    fn verify(&self, sig: S, msg: &str) -> bool;
+}
+
+/// A message signer using ECDSA
+pub struct EcdsaMessageSigner {
+    secp: Secp256k1<All>,
+    secret_key: SecretKey,
+}
+
+impl EcdsaMessageSigner {
+
+    /// Creates message signer from a bitcoin ECDSA private key
+    pub fn from_prv(prv: PrivateKey) -> Self {
+        Self::from_secret_key(prv.key)
+    }
+
+    /// Creates message signer from an ECDSA secret key
+    pub fn from_secret_key(secret_key: SecretKey) -> Self {
+        EcdsaMessageSigner {
+            secret_key,
+            secp: Secp256k1::new(),
+        }
+    }
+}
+
+impl MessageSigner<RecoverableSignature> for EcdsaMessageSigner {
+    fn sign(&self, message: &str) -> RecoverableSignature {
+        let msg_hash = signed_msg_hash(message);
+        self.secp.sign_recoverable(
+            &Message::from_slice(&msg_hash.into_inner()[..]).unwrap(),
```

Comment:
Either replace the unwrap with an expect (for reasonably unlikey errors) or change this function to return a `Result`